AsyncDisplayKit/ASMapNode.mm

```diff
@@ -0,0 +1,236 @@
+#import <AsyncDisplayKit/ASThread.h>
+
+@interface ASMapNode()
+{
+  ASDN::RecursiveMutex _propertyLock;
+  CGSize _nodeSize;
+  MKMapSnapshotter *_snapshotter;
+  MKMapSnapshotOptions *_options;
+  CGSize _maxSize;
+  NSArray *_annotations;
+}
+@end
+
+@implementation ASMapNode
+
+@synthesize liveMap = _liveMap;
+@synthesize mapSize = _mapSize;
+@synthesize automaticallyReloadsMapImageOnOrientationChange = _automaticallyReloadsMapImageOnOrientationChange;
+@synthesize mapDelegate = _mapDelegate;
+
+- (instancetype)initWithCoordinate:(CLLocationCoordinate2D)coordinate
+{
+  if (!(self = [super init])) {
+    return nil;
+  }
+  self.backgroundColor = ASDisplayNodeDefaultPlaceholderColor();
+  _automaticallyReloadsMapImageOnOrientationChange = YES;
+  _liveMap = NO;
+
+  _options = [[MKMapSnapshotOptions alloc] init];
+  _options.region = MKCoordinateRegionMakeWithDistance(coordinate, 1000, 1000);;
+  _mapImage = [[ASImageNode alloc]init];
+  _mapImage.clipsToBounds = YES;
+  [self addSubnode:_mapImage];
+  _maxSize = self.bounds.size;
+  return self;
+}
+
+- (void)addAnnotations:(NSArray *)annotations
+{
+  ASDN::MutexLocker l(_propertyLock);
+  if (annotations.count == 0) {
+    return;
+  }
+  _annotations = [annotations copy];
+  if (annotations.count != _annotations.count && _mapImage.image) {
+    // Redraw
+    [self setNeedsDisplay];
+  }
+}
+
+- (void)setUpSnapshotter
+{
+  if (!_snapshotter) {
+    _options.size = _nodeSize;
+    _snapshotter = [[MKMapSnapshotter alloc] initWithOptions:_options];
+  }
+}
+
+- (BOOL)isLiveMap
+{
+  ASDN::MutexLocker l(_propertyLock);
+  return _liveMap;
+}
+
+- (void)setLiveMap:(BOOL)liveMap
+{
+  ASDN::MutexLocker l(_propertyLock);
+  if (liveMap == _liveMap) return;
+  _liveMap = liveMap;
+  liveMap ? [self addLiveMap] : [self removeLiveMap];
+}
+
+- (CGSize)mapSize
+{
+  ASDN::MutexLocker l(_propertyLock);
+  return _mapSize;
+}
+
+- (void)setMapSize:(CGSize)mapSize
+{
+  ASDN::MutexLocker l(_propertyLock);
+  if (CGSizeEqualToSize(mapSize,_mapSize)) {
+    return;
+  }
+  _mapSize = mapSize;
+  _nodeSize = _mapSize;
+  _automaticallyReloadsMapImageOnOrientationChange = NO;
+  [self setNeedsLayout];
+}
+
+- (BOOL)automaticallyReloadsMapImageOnOrientationChange
+{
+  ASDN::MutexLocker l(_propertyLock);
+  return _automaticallyReloadsMapImageOnOrientationChange;
+}
+
+- (void)setAutomaticallyReloadsMapImageOnOrientationChange:(BOOL)automaticallyReloadsMapImageOnOrientationChange
+{
+  ASDN::MutexLocker l(_propertyLock);
+  if (_automaticallyReloadsMapImageOnOrientationChange == automaticallyReloadsMapImageOnOrientationChange) {
+    return;
+  }
+  _automaticallyReloadsMapImageOnOrientationChange = automaticallyReloadsMapImageOnOrientationChange;
+  
+}
+
+- (void)fetchData
+{
+  [super fetchData];
+  [self setUpSnapshotter];
+  [self takeSnapshot];
+}
+
+- (void)clearFetchedData
+{
+  [super clearFetchedData];
+  if (_mapView) {
+    [_mapView removeFromSupernode];
```

Comment:
To be honest this was what I was most unsure about but seemed to work in practice. Maybe there are better methods to call this code in? The current default behaviour is to revert back to the non-interactive map since if we didn't do this then the memory would spike, and it would be wasteful to have a MKMapView offscreen. I do like the idea of restoring the interactive map, if _liveMap = YES, sort of like autoplaying videos in FB. I think we should take this route because either the developer or the user has interacted in a way to turn the snapshot into a interactive map, so it would make sense to recreate the same experience that they last saw.